### ISSUE TITLE
Massive descrepancy between Diffusers results and results from these nodes

### ISSUE THREAD

**Issue by Astropulse** [ExponentialML/ComfyUI_ELLA#14]
I'm not sure what is causing it, but this clearly is not working properly.

Diffusers gives the expected result, but this repo gives worse results than normal clip guidance.

Diffusers vs this node
![image](https://github.com/ExponentialML/ComfyUI_ELLA/assets/61034487/cb8608d6-9791-4cfd-9a65-2da65461eb3f)

Regular old clip
![ComfyUI_05734_](https://github.com/ExponentialML/ComfyUI_ELLA/assets/61034487/ea500874-8481-481c-8d45-e7cf8d40c3b3)

Has anyone been able to get the expected results using this?

**Comment by renwerx** [comment]
I was going to observe that you hadn't set your sigma but then I tried and my results are even worse, it seems to fundamentally change the style of the model. This is a comparison for the prompt:

A smiling friendly elderly (Catholic priest:1) from 1850s Germany, wearing glasses and a clerical collar, a crowd of factory labourers in the background

Vanilla CLIP:
![20240412_CLIP](https://github.com/ExponentialML/ComfyUI_ELLA/assets/132638531/af3912a6-640d-4a4e-ba98-335ed6a69fee)

This repo:
![20240412_ELLA](https://github.com/ExponentialML/ComfyUI_ELLA/assets/132638531/cd43f8d9-267b-4476-9564-8183b4707966)

**Comment by Astropulse** [comment]
Yeah even setting the sigma properly results in unexpected results. ELLA through diffusers mitigates the color token bleed expected in clip, while ELLA from this repo does not. Results across more tests give similar results, with ELLA from both methods being better than clip, but the results from this repo being clearly inferior either in accuracy or image quality...
![image](https://github.com/ExponentialML/ComfyUI_ELLA/assets/61034487/10ff4c6d-b1f4-4bfb-8528-8a8882ee5454)

**Comment by Creative-comfyUI** [comment]
The problem came from the scheduler and the sampler. Also, some models are much more accurate than others. Be sure it is sd 1.5. Try using LCM and sgm_uniform, CFG 4, step 40. Don't forget that every model has its own specification. This can also change the results. Some models add color information to other parts of the image. I have tested the examples on this page and it works perfectly. The CFG depends sometimes on the model, some are better with a small number than a larger one. It seems that when there is a lot of objects we have less accurate results.

**Comment by Astropulse** [comment]
These are not solutions to the issue, as it is working properly on nearly all models via diffusers. Changing the sampler or scheduler does not improve results, as this is a conditioning related issue, not a sampling issue. You can clearly see that the created conditioning is not correct/in line with the proper implementation in diffusers.

**Comment by Creative-comfyUI** [comment]
![Screenshot 2024-04-12 at 2 44 26 PM](https://github.com/ExponentialML/ComfyUI_ELLA/assets/166729777/c0a08a99-0d54-4d72-8020-cbb4c547ac5c)

**Comment by Astropulse** [comment]
There is still very clearly color bleed from the embedding. With a proper implementation of ella this should not be happening (the green hair)
![image](https://github.com/ExponentialML/ComfyUI_ELLA/assets/61034487/8ec50dcb-55bb-4f7d-98e4-d70a773215c4)

Setting the get sigma node step values to the same as the sampler actually make this even worse:
![image](https://github.com/ExponentialML/ComfyUI_ELLA/assets/61034487/a71a87d2-6067-4236-a2f8-a1534e4066a7)

You can see issues exaggerated if you increase the cfg value. (Also worse when sigma steps are the same as sampling)
![image](https://github.com/ExponentialML/ComfyUI_ELLA/assets/61034487/628c2eb4-3040-4a84-84df-21605dcefe4a)

Like I keep saying, this is not a sampling issue. (ELLA and this model should work fine with euler, or any other sampler anyways)
My best guess is that the get sigmas node is not returning the correct values, or there was a mistake porting over the ELLA code that I cant seem to find.

Bottom line is that there IS an issue. Diffusers is the official, expected result. Other implementations should not have deviation from it in terms of sampler compatibility, if they do it indicates a mistake somewhere in the chain. Please actually look into this instead of dismissing my observations.

**Comment by Creative-comfyUI** [comment]
First of all, I am not dismissing your observations, but I did not understand your point of view. Now it is clearer. 

How do you explain that we get better or more consistent results for some models? I have seen this problem even with basic nodes, often with SD 1-1.5 models. For the same prompt some models don't give the same result. 

The other ella node does not work on my side, it is so slow that I cannot do anything else on my computer at the same time.

**Comment by Astropulse** [comment]
Thanks, the color assignment is a symptom of the embedding not being created properly, with tokens influencing other areas of the embedding. It's just a good way of testing if something is behaving as expected with ELLA and T5 XL. Likely the reason you cant run the other node is because of T5 XL and SD being loaded at the same time because diffusers doesnt optimize things well.

Different results can be explained by ELLA using a fundamentally different text encoder space, so some variance is expected, especially with style shifts as seen in the second post in this issue.

A main thing to keep in mind is that because ELLA replaces the *default* clip embeddings, only models with largely un-modified clip sections can be used. This means models that incorporate the NAI leak, or heavily train the text encoder will not give good results. (This is why I'm using Dreamshaper version 8, as its clip space is mostly intact) Some training on the text encoder is fine, but if its too far then ella wont be able to create the embeddings the model's unet is expecting.

I still haven't been able to find where this port went wrong... but thats probably partially because I have a really hard time understanding diffusers code at all. I'm trying to compare values of the tensors and embeddings across timesteps but that isnt showing anything blatantly strange.

**Comment by quixot1c** [comment]
Same issue: https://github.com/ExponentialML/ComfyUI_ELLA/issues/12. Linked for repo owner bookkeeping.

**Comment by kijai** [comment]
I think ELLA itself expects much higher value for the timestep than what this method of using the sigma gives, simply giving it a value like 800 makes it follow the prompt in similar fashion than the original code:
![image](https://github.com/ExponentialML/ComfyUI_ELLA/assets/40791699/d8744b54-7528-46a4-bdff-cae02405c1ea)

**Comment by kijai** [comment]
Made a PR about this now, would appreciate help testing it, so far it's much closer to the original results. 
![image](https://github.com/ExponentialML/ComfyUI_ELLA/assets/40791699/f0c5d9e0-c261-46d8-8000-4c379ce9cde6)

**Comment by Astropulse** [comment]
Might be a difference in how diffusers handles timesteps. Appreciate your help here, and I'll test out the PR.

**Comment by samcroft** [comment]
> Made a PR about this now, would appreciate help testing it, so far it's much closer to the original results. 

I've been testing this out and it works great 👍

Could you explain what adjusting the manual timestamp value in the ELLA Text Encode prompt node does?

**Comment by kijai** [comment]
> > Made a PR about this now, would appreciate help testing it, so far it's much closer to the original results.
> 
> I've been testing this out and it works great 👍
> 
> Could you explain what adjusting the manual timestamp value in the ELLA Text Encode prompt node does?

I left it for debugging purposes mostly, if you don't provide the model and sigmas, it's used instead. If you use a low value, like the GetSigmas node provides (around 80) you get similar results as before. Using something in range of 600-999 starts giving more prompt adherence, but also affects image quality. This is only the value fed to ELLA itself, and it definitely expects the values as the actual timesteps (1000-0 range).

When using the model and sigmas as input, my method is to calculate the corresponding timestep for each sigma and process the cond with ELLA, then map those conds to comfy conds using the set start/end percentage so they are used similarly as in the diffusers implementation.

I don't know if there's some downside in doing it like this, feels a bit hacky to me, but I haven't run into any issues yet,  sampling speed didn't change and results improved. Possibly a more proper way to do it could be as model patch instead, that's beyond my current level of expertise though.

**Comment by renwerx** [comment]
I've had some time to play around with the PR and I'm super happy so far, even if it's not "ideal" it makes this repo very useful indeed. Thanks guys!

**Comment by kijai** [comment]
Tencent has released official nodes now too: https://github.com/TencentQQGYLab/ComfyUI-ELLA
